cpp/src/arrow/flight/client.cc

```diff
@@ -259,6 +111,65 @@ class FlightStreamReader : public RecordBatchReader {
   std::unique_ptr<grpc::ClientReader<pb::FlightData>> stream_;
 };
 
+class FlightClient;
+
+/// \brief A RecordBatchWriter implementation that writes to a Flight
+/// DoPut stream.
+class FlightStreamWriter : public ipc::RecordBatchWriter {
+ public:
+  explicit FlightStreamWriter(std::unique_ptr<ClientRpc>&& rpc,
+                              const FlightDescriptor& descriptor,
+                              const std::shared_ptr<Schema>& schema)
+      : rpc_{std::move(rpc)},
+        descriptor_{descriptor},
+        schema_{schema},
+        pool_{default_memory_pool()} {}
+
+  Status WriteRecordBatch(const RecordBatch& batch, bool allow_64bit = false) override {
+    IpcPayload payload;
+    RETURN_NOT_OK(ipc::internal::GetRecordBatchPayload(batch, pool_, &payload));
+    auto custom_writer = reinterpret_cast<grpc::ClientWriter<IpcPayload>*>(writer_.get());
+    // Explicitly specify the override to invoke - otherwise compiler
+    // may invoke through vtable (not updated by reinterpret_cast)
+    if (!custom_writer->grpc::ClientWriter<IpcPayload>::Write(payload,
+                                                              grpc::WriteOptions())) {
+      // Stream ended?
+      return Status::UnknownError("Could not write record batch to stream");
+    }
+    return Status::OK();
+  }
+
+  Status Close() override {
+    bool finished_writes = writer_->WritesDone();
+    RETURN_NOT_OK(internal::FromGrpcStatus(writer_->Finish()));
+    if (!finished_writes) {
+      return Status::UnknownError(
+          "Could not finish writing record batches before closing");
+    }
+    return Status::OK();
+  }
+
+  void set_memory_pool(MemoryPool* pool) override { pool_ = pool; }
+
+ private:
+  /// \brief Set the gRPC writer backing this Flight stream.
+  /// \param [in] writer the gRPC writer
+  void set_stream(std::unique_ptr<grpc::ClientWriter<pb::FlightData>>&& writer) {
+    writer_ = std::move(writer);
+  }
+
+  // TODO: there isn't a way to access this as a user.
```

Comment:
As I was reading this I was thinking about whether returning `arrow::ipc::RecordBatchWriter` is the best thing. If we need other kinds of methods related to acks or RPC-specific details, it might be worth defining an Flight-specific class that is returned from these methods

Reply:
I was debating this as well. It should at least be easy to wrap a custom interface in `RecordBatchWriter`, so I guess it's better to leave this open for future extensions.

Reply:
I turned this into a pImpl, the wrapper doesn't expose this still, but leaves it open for the future. The wrapper is still a RecordBatchWriter; should it be further decoupled?

Reply:
For now I think we can keep it as a `RecordBatchWriter`, especially as the base class is just an interface.
Furthermore, we probably shouldn't consider the Flight API stable until it's battle-tested.

Reply:
I've changed it back to RecordBatchWriter.